### PR TITLE
Added unit tests for Legacy Notification Builder 

### PR DESCRIPTION
**Pull request by PratthamArora**
## Description
Added unit tests for Legacy Notification Builder 

## Related Issue
https://jira.corp.adobe.com/browse/MOB-20957

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+class LegacyNotificationBuilderTest {
+
+    @Mock
+    private lateinit var pushTemplate: BasicPushTemplate
+    private lateinit var trackerActivityClass: Class<out Activity>
+    private lateinit var context: Context
+    private val SUT = LegacyNotificationBuilder
+
+    @Before
+    fun setUp() {
```
Comment (prudrabhat) [inline code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:49]: IMO creating tests that ultimately check if a notification has been constructed(or not) as desired with all the values present/absent would be more robust as opposed to individually asserting on a single setter at a time. I feel verifying individual keys will make it cumbersome to test interrelated  keys/settings later on.

I am wondering if we have the following setup 
-  a map/bundle with all the keys available for the template
- an extension method on this map/bundle that allows removing/unsetting a list of values/keys.
- an extension method on this map/bundle that allows overriding a list of keys via an input map

it will enable us to write tests in the format of 

```
val myTemplateMap = getLegacyTemplateDataMap()  // returns fully populated template with all test data
                              .without(list(KEY1, KEY7,....)) // some keys with additional clauses/defaults in source code
                              .overwrite(mapOf(.....))
val mapData = MapData(myTemplateMap)
LegacyNotificationBuilder.construct(..., mapData,....)
verifyConstruction(mapData, notification, omittedKeys, overwrittenKeys)  // intelligent method to compare at notification data and built notification
```

@PratthamArora , @spoorthipujariadobe , @PravinPK - Thoughts?

Comment (spoorthipujariadobe) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:49]: Agree with your suggestion. This PR has the utility methods you're referring to #9. Should be simpler to use those to check if the notification was constructed properly for given data. You can add/remove keys or replace values to test the various code branches we have. Since these are unit tests, we can adopt a white box approach

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [31])
+class LegacyNotificationBuilderTest {
+
+    @Mock
+    private lateinit var pushTemplate: BasicPushTemplate
+    private lateinit var trackerActivityClass: Class<out Activity>
+    private lateinit var context: Context
+    private val SUT = LegacyNotificationBuilder
```
[the remark this answers is not in the document]
Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:46]: sure, will use the singleton directly

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt (hunk 1 above)
Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:49]: sure, will create a mock data class with dummy data and merge all the happy cases for the construct method

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            LegacyNotificationBuilder.construct(context, pushTemplate, trackerActivityClass)
+        val pendingIntent = notificationBuilder.build().contentIntent
+        val shadowPendingIntent = Shadows.shadowOf(pendingIntent)
+        val intent = shadowPendingIntent.savedIntent
+
+        assertEquals(NotificationCompat.Builder::class.java, notificationBuilder.javaClass)
+        assertEquals(pushTemplate.ticker, notificationBuilder.build().tickerText)
+        assertEquals(
+            pushTemplate.title,
+            notificationBuilder.build().extras.getString(NotificationCompat.EXTRA_TITLE)
```
Comment (prudrabhat) [inline code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:70]: Can you store the built notification in a variable instead of calling `notificationBuilder.build()` every time it is needed. Or, is there a limitation that needs this approach?

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/MockDataUtils.kt [hunk 4]
```diff
@@ -20,7 +20,7 @@ const val MOCKED_BODY = "Mocked Body"
 const val MOCKED_PAYLOAD_VERSION = "1"
 const val MOCKED_CAROUSEL_LAYOUT = "default"
 const val MOCKED_BODY_TEXT_COLOR = "#FFFFFF"
-const val MOCKED_SMALL_ICON = "chat_bubble"
+const val MOCKED_SMALL_ICON = "skipleft"
```
Comment (prudrabhat) [inline code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/MockDataUtils.kt:23]: Is this change intentional? Does this refer to an asset?

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        assertEquals(
+            pushTemplate.tag,
+            intent.getStringExtra(PushTemplateConstants.PushPayloadKeys.TAG)
+        )
+        assertEquals(
+            pushTemplate.isNotificationSticky,
+            intent.getBooleanExtra(PushTemplateConstants.PushPayloadKeys.STICKY, false)
+        )
+    }
+
```
Comment (prudrabhat) [inline code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:100]: For these test cases below which verify specific condtions, can you also verify that the other attributes of the notification are set as desired?

**Review by prudrabhat — CHANGES_REQUESTED** [review]

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/MockDataUtils.kt (hunk 4 above)
Comment (navratan-soni) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/MockDataUtils.kt:23]: Yes this is an intentional change.

Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/MockDataUtils.kt:23]: we need this since this asset is available in the resource directory

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt (hunk 3 above)
Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:70]: done

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt (hunk 5 above)
Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:100]: have added a test case to verify sticky notification values but apart from that it is only being used to create a pending intent

**Review by navratan-soni — APPROVED** [review]
Looks good to merge !
Approving as all the review comments are addressed.

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt (hunk 5 above)
Comment (prudrabhat) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:102]: Sounds good.  
My suggestion was more on the lines of asserting that changing or modifying data for the test cases below does not affect other attributes. This approach will ensure that the test case breaks if any linked breaking changes (channel id somehow modifying priority or vibration status) are done later during development forcing the developer to add or modify tests per their new requirements.

```
fun `construct should set default channel ID if pushTemplate channelId is null`() {
   // 1.  Remove key from mock data
   dataMap.removeKeysFromMap(PushTemplateConstants.PushPayloadKeys.CHANNEL_ID)
   
  // 2.  verify outcome of that specific key
  assertEquals(DEFAULT_CHANNEL_ID, notification.channelId)

   // 3. Verify that rest of the notification attributes are same as the ones derived from mock data (i.e and that this does not affect anything else).
 
}
```

**Review by prudrabhat — APPROVED** [review]
Added a suggestion on ensuring test case completeness.

File: code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt (hunk 5 above)
Comment (PratthamArora) [reply code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/LegacyNotificationBuilderTest.kt:102]: sure, will update the test case for Default channel ID to verify other notification attributes as well
